src/realm.h

```diff
@@ -4069,17 +4089,93 @@ RLM_API realm_sync_socket_t* realm_sync_socket_new(
     realm_sync_socket_websocket_async_write_func_t websocket_write_func,
     realm_sync_socket_websocket_free_func_t websocket_free_func);
 
-RLM_API void realm_sync_socket_callback_complete(realm_sync_socket_callback_t* realm_callback, realm_errno_e status,
-                                                 const char* reason);
+/**
+ * This function is to be called by the CAPI Implementer when the timer expires or an error occurs.
+ * @param timer_handler the timer callback handler that was provided when the timer was created.
+ * @param status the error code for the error that occurred or RLM_ERR_NONE if the timer expired normally.
+ * @param reason a string describing details about the error that occurred or empty string if no error.
+ * NOTE: This function must be called by the event loop execution thread.
+ */
+RLM_API void realm_sync_socket_timer_complete(realm_sync_socket_callback_t* timer_handler, realm_errno_e status,
```

Comment:
aren't these two new api calls leaking the callback if realm_sync_socket_post_complete is never called?

Reply:
Maybe I should rethink the callback handler a bit differently so the callback isn't called multiple times...

Reply:
Yeah, that's an option. I just saw you kept the release in dtor, so there is no leak. But, how is double free gone? (can't the callback be invoked followed by release, and then when the timer is gone there is a new release?)

Reply:
The double free is gone, since you're supposed to use the `timer_complete()` or `timer_canceled()` fcns, which doesn't call release, so there's only one release when the timer is destroyed (if the CAPI is implemented properly), however, the callback could be called multiple times depending on how the timer is implemented. (cancel is always called in the destructor to make sure the timer is stopped)

Reply:
but `post` calls release..

Reply:
But you're not supposed to use post (or write_complete) with the timer handler, which are different objects from the callback handlers provided to the post or async_write functions.